src/poetry/core/packages/constraints/multi_constraint.py

```diff
@@ -71,18 +66,20 @@ def allows_any(self, other: "ConstraintTypes") -> bool:
 
         return False
 
-    def intersect(self, other: Constraint) -> "MultiConstraint":
-        if isinstance(other, Constraint):
-            constraints = self._constraints
-            if other not in constraints:
-                constraints += (other,)
-            else:
-                constraints = (other,)
+    def intersect(self, other: BaseConstraint) -> "BaseConstraint":
+        if not isinstance(other, Constraint):
```

Comment:
If we only allow `Constraint` we should tell it so by the type hint.

Reply:
But then this signature would definitely be inconsistent with the parent signature, which mypy would rightly complain about.

It is logically allowed to provide a `Constraint` here, it's just that code to support this case has not been written.  (And, presumably, it does not arise in practice.)